src/audio/copier/copier_generic.c

```diff
@@ -274,10 +274,39 @@ pcm_converter_func get_converter_func(const struct ipc4_audio_format *in_fmt,
 	audio_stream_fmt_conversion(out_fmt->depth, out_fmt->valid_bit_depth, &out, &out_valid,
 				    out_fmt->s_type);
 
-	if (in_fmt->s_type == IPC4_TYPE_MSB_INTEGER && in_valid == SOF_IPC_FRAME_S24_4LE)
-		in_valid = SOF_IPC_FRAME_S24_4LE_MSB;
-	if (out_fmt->s_type == IPC4_TYPE_MSB_INTEGER && out_valid == SOF_IPC_FRAME_S24_4LE)
-		out_valid = SOF_IPC_FRAME_S24_4LE_MSB;
+	if (in_fmt->s_type == IPC4_TYPE_MSB_INTEGER && in_valid == SOF_IPC_FRAME_S24_4LE) {
+		switch (type) {
+		case ipc4_gtw_host:
+			if (dir == ipc4_playback)
+				in_valid = SOF_IPC_FRAME_S24_4LE_MSB;
+			break;
+		case ipc4_gtw_alh:
+		case ipc4_gtw_link:
+		case ipc4_gtw_ssp:
+			if (dir == ipc4_capture)
+				in_valid = SOF_IPC_FRAME_S24_4LE_MSB;
+			break;
+		default:
+			break;
+		}
+	}
+
+	if (out_fmt->s_type == IPC4_TYPE_MSB_INTEGER && out_valid == SOF_IPC_FRAME_S24_4LE) {
+		switch (type) {
+		case ipc4_gtw_host:
+			if (dir == ipc4_capture)
+				out_valid = SOF_IPC_FRAME_S24_4LE_MSB;
+			break;
+		case ipc4_gtw_alh:
+		case ipc4_gtw_link:
+		case ipc4_gtw_ssp:
```

Comment:
missing gtw_dmic type

Reply:
Thanks, if you have validate all the cases, I can start to refine it now